File: modules/session-audio-engine/ios/PendingCaptureStore.swift
import Foundation

final class PendingCaptureStore {
  private let fileManager = FileManager.default
  private let encoder = JSONEncoder()
  private let decoder = JSONDecoder()
  private let manifestURL: URL
  private var segments: [NativeCapturedSegment]

  init() {
    let support = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
    let directory = support.appendingPathComponent("session-audio-engine", isDirectory: true)
    try? fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
    manifestURL = directory.appendingPathComponent("pending-captures.json")
    if let data = try? Data(contentsOf: manifestURL),
       let decoded = try? decoder.decode([NativeCapturedSegment].self, from: data) {
      segments = decoded
    } else {
      segments = []
    }
    let documents = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
    let captureDirectory = documents
      .appendingPathComponent("recordings", isDirectory: true)
      .appendingPathComponent("session-captures", isDirectory: true)
    try? fileManager.createDirectory(at: captureDirectory, withIntermediateDirectories: true)
    segments = segments.compactMap { segment in
      let finalURL = captureDirectory.appendingPathComponent(segment.fileName)
      let temporaryURL = captureDirectory.appendingPathComponent(".\(segment.fileName).tmp")
      if !fileManager.fileExists(atPath: finalURL.path), fileManager.fileExists(atPath: temporaryURL.path) {
        try? fileManager.moveItem(at: temporaryURL, to: finalURL)
      }
      guard fileManager.fileExists(atPath: finalURL.path) else { return nil }
      return NativeCapturedSegment(
        segmentId: segment.segmentId,
        sessionId: segment.sessionId,
        uri: finalURL.absoluteString,
        fileName: segment.fileName,
        phase: segment.phase,
        cycle: segment.cycle,
        capturedAt: segment.capturedAt,
        durationMs: segment.durationMs,
        speechStartMs: segment.speechStartMs,
        speechEndMs: segment.speechEndMs
      )
    }
    try? persist()
  }

  func store(
    samples: [Int16],
    configuration: NativeSessionConfiguration,
    phase: String,
    cycle: Int,
    speechStartMs: Int64,
    speechEndMs: Int64
  ) throws -> NativeCapturedSegment {
    guard let directoryURL = URL(string: configuration.captureDirectoryUri) else {
      throw SessionAudioEngineError.storageUnavailable
    }
    let segmentId = UUID().uuidString.lowercased()
    let fileName = "session-\(configuration.sessionId)-\(segmentId).wav"
    let temporaryURL = directoryURL.appendingPathComponent(".\(fileName).tmp")
    let finalURL = directoryURL.appendingPathComponent(fileName)
    let durationMs = Int64(samples.count) * 1000 / 16_000
    let segment = NativeCapturedSegment(
      segmentId: segmentId,
      sessionId: configuration.sessionId,
      uri: finalURL.absoluteString,
      fileName: fileName,
      phase: phase,
      cycle: cycle,
      capturedAt: ISO8601DateFormatter().string(from: Date()),
      durationMs: durationMs,
      speechStartMs: max(0, min(speechStartMs, durationMs)),
      speechEndMs: max(0, min(speechEndMs, durationMs))
    )
    let wav = makeWavData(samples: samples)
    try wav.write(to: temporaryURL, options: .atomic)
    segments.append(segment)
    try persist()
    try fileManager.moveItem(at: temporaryURL, to: finalURL)
    evictOverLimit(configuration.maxPendingCaptureBytes)
    try persist()
    return segment
  }

  func all() -> [NativeCapturedSegment] {
    segments.filter { fileManager.fileExists(atPath: URL(string: $0.uri)?.path ?? "") }
  }

  func markStored(ids: Set<String>) throws {
    segments.removeAll { ids.contains($0.segmentId) }
    try persist()
  }

  private func evictOverLimit(_ limit: Int64) {
    guard limit > 0 else { return }
    var total = segments.reduce(Int64(0)) { partial, segment in
      let size = (try? fileManager.attributesOfItem(atPath: URL(string: segment.uri)?.path ?? "")[.size] as? NSNumber)?.int64Value ?? 0
      return partial + size
    }
    while total > limit, let oldest = segments.first {
      let url = URL(string: oldest.uri)
      let size = (try? fileManager.attributesOfItem(atPath: url?.path ?? "")[.size] as? NSNumber)?.int64Value ?? 0
      if let url { try? fileManager.removeItem(at: url) }
      segments.removeFirst()
      total -= size
    }
  }

  private func persist() throws {
    try encoder.encode(segments).write(to: manifestURL, options: .atomic)
  }

  private func makeWavData(samples: [Int16]) -> Data {
    let dataSize = UInt32(samples.count * MemoryLayout<Int16>.size)
    var data = Data()
    data.append(contentsOf: Array("RIFF".utf8))
    data.appendLittleEndian(UInt32(36) + dataSize)
    data.append(contentsOf: Array("WAVEfmt ".utf8))
    data.appendLittleEndian(UInt32(16))
    data.appendLittleEndian(UInt16(1))
    data.appendLittleEndian(UInt16(1))
    data.appendLittleEndian(UInt32(16_000))
    data.appendLittleEndian(UInt32(32_000))
    data.appendLittleEndian(UInt16(2))
    data.appendLittleEndian(UInt16(16))
    data.append(contentsOf: Array("data".utf8))
    data.appendLittleEndian(dataSize)
    samples.forEach { data.appendLittleEndian(UInt16(bitPattern: $0)) }
    return data
  }
}

private extension Data {
  mutating func appendLittleEndian<T: FixedWidthInteger>(_ value: T) {
    var littleEndian = value.littleEndian
    Swift.withUnsafeBytes(of: &littleEndian) { append(contentsOf: $0) }
  }
}

File: modules/session-audio-engine/android/src/main/java/com/joynnovate/buddybird/sessionaudio/PendingCaptureStore.kt
package com.joynnovate.buddybird.sessionaudio

import android.content.Context
import android.net.Uri
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedOutputStream
import java.io.DataOutputStream
import java.io.File
import java.io.FileOutputStream
import java.time.Instant
import java.util.UUID

class PendingCaptureStore(context: Context) {
  private val directory = File(context.filesDir, "session-audio-engine").apply { mkdirs() }
  private val manifest = File(directory, "pending-captures.json")
  private val segments = load().toMutableList()

  @Synchronized
  fun reconcile(captureDirectoryUri: String) {
    val captureDirectory = File(requireNotNull(Uri.parse(captureDirectoryUri).path)).apply { mkdirs() }
    val registeredTemporaryFiles = segments.mapTo(mutableSetOf()) { ".${it.fileName}.tmp" }
    captureDirectory.listFiles()
      ?.filter { it.isFile && it.name.startsWith(".") && it.name.endsWith(".wav.tmp") }
      ?.filterNot { it.name in registeredTemporaryFiles }
      ?.forEach(File::delete)
  }

  @Synchronized
  fun store(
    samples: ShortArray,
    configuration: NativeSessionConfiguration,
    phase: String,
    cycle: Int,
    speechStartMs: Long,
    speechEndMs: Long,
  ): NativeCapturedSegment {
    val captureDirectory = File(requireNotNull(Uri.parse(configuration.captureDirectoryUri).path)).apply { mkdirs() }
    val segmentId = UUID.randomUUID().toString().lowercase()
    val fileName = "session-${configuration.sessionId}-$segmentId.wav"
    val temporary = File(captureDirectory, ".$fileName.tmp")
    val final = File(captureDirectory, fileName)
    val durationMs = samples.size.toLong() * 1_000 / 16_000
    val segment = NativeCapturedSegment(
      segmentId = segmentId,
      sessionId = configuration.sessionId,
      uri = Uri.fromFile(final).toString(),
      fileName = fileName,
      phase = phase,
      cycle = cycle,
      capturedAt = Instant.now().toString(),
      durationMs = durationMs,
      speechStartMs = speechStartMs.coerceIn(0, durationMs),
      speechEndMs = speechEndMs.coerceIn(0, durationMs),
    )
    writeWav(temporary, samples)
    segments += segment
    persist()
    check(temporary.renameTo(final)) { "Could not finalize captured WAV file." }
    evictOverLimit(configuration.maxPendingCaptureBytes)
    persist()
    return segment
  }

  @Synchronized fun all(): List<NativeCapturedSegment> = segments.filter { File(requireNotNull(Uri.parse(it.uri).path)).isFile }

  @Synchronized
  fun markStored(ids: Set<String>) {
    segments.removeAll { ids.contains(it.segmentId) }
    persist()
  }

  private fun load(): List<NativeCapturedSegment> {
    if (!manifest.isFile) return emptyList()
    return runCatching {
      val array = JSONArray(manifest.readText())
      buildList {
        for (index in 0 until array.length()) {
          val value = array.getJSONObject(index)
          val segment = NativeCapturedSegment(
            segmentId = value.getString("segmentId"),
            sessionId = value.getString("sessionId"),
            uri = value.getString("uri"),
            fileName = value.getString("fileName"),
            phase = value.getString("phase"),
            cycle = value.getInt("cycle"),
            capturedAt = value.getString("capturedAt"),
            durationMs = value.getLong("durationMs"),
            speechStartMs = value.getLong("speechStartMs"),
            speechEndMs = value.getLong("speechEndMs"),
          )
          val final = File(requireNotNull(Uri.parse(segment.uri).path))
          val temporary = File(final.parentFile, ".${segment.fileName}.tmp")
          if (!final.exists() && temporary.exists()) temporary.renameTo(final)
          if (final.isFile) add(segment)
        }
      }
    }.getOrDefault(emptyList())
  }

  private fun persist() {
    val array = JSONArray()
    segments.forEach { array.put(JSONObject(it.toMap())) }
    val temporary = File(directory, ".pending-captures.json.tmp")
    temporary.writeText(array.toString())
    check(temporary.renameTo(manifest)) { "Could not finalize pending capture manifest." }
  }

  private fun evictOverLimit(limit: Long) {
    if (limit <= 0) return
    var total = segments.sumOf { File(requireNotNull(Uri.parse(it.uri).path)).length() }
    while (total > limit && segments.isNotEmpty()) {
      val oldest = segments.removeAt(0)
      val file = File(requireNotNull(Uri.parse(oldest.uri).path))
      total -= file.length()
      file.delete()
    }
  }

  private fun writeWav(file: File, samples: ShortArray) {
    DataOutputStream(BufferedOutputStream(FileOutputStream(file))).use { output ->
      val dataSize = samples.size * 2
      output.writeBytes("RIFF")
      output.writeIntLE(36 + dataSize)
      output.writeBytes("WAVEfmt ")
      output.writeIntLE(16)
      output.writeShortLE(1)
      output.writeShortLE(1)
      output.writeIntLE(16_000)
      output.writeIntLE(32_000)
      output.writeShortLE(2)
      output.writeShortLE(16)
      output.writeBytes("data")
      output.writeIntLE(dataSize)
      samples.forEach { output.writeShortLE(it.toInt()) }
    }
  }
}

private fun DataOutputStream.writeIntLE(value: Int) {
  writeByte(value)
  writeByte(value ushr 8)
  writeByte(value ushr 16)
  writeByte(value ushr 24)
}

private fun DataOutputStream.writeShortLE(value: Int) {
  writeByte(value)
  writeByte(value ushr 8)
}
